File: go/internal/ws/graphql_ws.go
package ws

import (
	"encoding/json"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/SidathWeerasinghe/mock-services/go/internal/graphql"
	payloadgen "github.com/SidathWeerasinghe/mock-services/go/internal/payload"
	"github.com/gorilla/websocket"
)

const minIntervalMs = 500

func GraphQLWS(w http.ResponseWriter, r *http.Request) {
	conn, err := Upgrader.Upgrade(w, r, nil)
	if err != nil {
		return
	}
	defer conn.Close()

	subs := make(map[string]chan struct{})
	var mu sync.Mutex

	stopAll := func() {
		mu.Lock()
		defer mu.Unlock()
		for id, ch := range subs {
			close(ch)
			delete(subs, id)
		}
	}
	defer stopAll()

	send := func(v interface{}) {
		b, _ := json.Marshal(v)
		_ = conn.WriteMessage(websocket.TextMessage, b)
	}

	for {
		_, raw, err := conn.ReadMessage()
		if err != nil {
			break
		}
		var msg map[string]interface{}
		if err := json.Unmarshal(raw, &msg); err != nil {
			send(map[string]interface{}{"type": "error", "payload": []map[string]string{{"message": "Invalid JSON frame"}}})
			continue
		}
		msgType, _ := msg["type"].(string)
		subID, _ := msg["id"].(string)

		switch msgType {
		case "connection_init":
			send(map[string]string{"type": "connection_ack"})

		case "subscribe":
			subPayload, _ := msg["payload"].(map[string]interface{})
			vars, _ := subPayload["variables"].(map[string]interface{})
			query, _ := subPayload["query"].(string)
			if !strings.Contains(strings.ToLower(strings.ReplaceAll(query, "_", "")), "mockstream") {
				send(map[string]interface{}{
					"id": subID, "type": "error",
					"payload": []map[string]string{{"message": "Only mockStream subscription is supported"}},
				})
				continue
			}
			if vars == nil {
				vars = graphql.ParseInlineVars(query, nil)
			}
			resource := graphqlStrVar(vars, "resource", "events")
			size := graphqlIntVar(vars, "size", 1)
			format := graphqlStrVar(vars, "format", "json")
			method := graphqlStrVar(vars, "method", "GET")
			intervalMs := graphqlIntVar(vars, "intervalMs", 1000)
			if intervalMs < minIntervalMs {
				intervalMs = minIntervalMs
			}

			mu.Lock()
			if ch, ok := subs[subID]; ok {
				close(ch)
			}
			stop := make(chan struct{})
			subs[subID] = stop
			mu.Unlock()

			go func(id string, stopCh chan struct{}) {
				emit := func() bool {
					data, err := graphql.MakeMockStreamResponse(method, resource, size, format)
					if err != nil {
						send(map[string]interface{}{
							"id": id, "type": "error",
							"payload": []map[string]interface{}{{"message": err.Error(), "validSizes": payloadgen.ValidSizesKB}},
						})
						return false
					}
					send(map[string]interface{}{
						"id": id, "type": "next",
						"payload": map[string]interface{}{"data": map[string]interface{}{"mockStream": data}},
					})
					return true
				}
				if !emit() {
					return
				}
				ticker := time.NewTicker(time.Duration(intervalMs) * time.Millisecond)
				defer ticker.Stop()
				for {
					select {
					case <-stopCh:
						return
					case <-ticker.C:
						if !emit() {
							return
						}
					}
				}
			}(subID, stop)

		case "complete":
			mu.Lock()
			if ch, ok := subs[subID]; ok {
				close(ch)
				delete(subs, subID)
			}
			mu.Unlock()
			send(map[string]string{"id": subID, "type": "complete"})

		case "ping":
			send(map[string]string{"type": "pong"})
		}
	}
}

func graphqlStrVar(vars map[string]interface{}, key, def string) string {
	if v, ok := vars[key]; ok {
		switch t := v.(type) {
		case string:
			return t
		default:
			b, _ := json.Marshal(t)
			return strings.Trim(string(b), `"`)
		}
	}
	return def
}

func graphqlIntVar(vars map[string]interface{}, key string, def int) int {
	if v, ok := vars[key]; ok {
		switch t := v.(type) {
		case float64:
			return int(t)
		case int:
			return t
		}
	}
	return def
}

File: nodejs/src/ws/graphqlWs.js
'use strict';

const { WebSocketServer } = require('ws');
const { makeMockStreamResponse, VALID_SIZES_KB, parseInlineVars } = require('../graphql/executor');

const MIN_INTERVAL_MS = 500;

function attachGraphqlWs(server, path = '/graphql-ws') {
  const wss = new WebSocketServer({ server, path });

  wss.on('connection', (ws) => {
    const subs = new Map();
    const send = (obj) => ws.send(JSON.stringify(obj));

    ws.on('message', (raw) => {
      let msg;
      try {
        msg = JSON.parse(raw.toString());
      } catch {
        send({ type: 'error', payload: [{ message: 'Invalid JSON frame' }] });
        return;
      }

      const type = msg.type;
      const id = String(msg.id || '');

      if (type === 'connection_init') {
        send({ type: 'connection_ack' });
        return;
      }

      if (type === 'subscribe') {
        const payload = msg.payload || {};
        let variables = payload.variables || {};
        const query = (payload.query || '').toLowerCase();
        if (!query.replace(/_/g, '').includes('mockstream')) {
          send({ id, type: 'error', payload: [{ message: 'Only mockStream subscription is supported' }] });
          return;
        }
        variables = parseInlineVars(payload.query || '', variables);

        if (subs.has(id)) {
          clearInterval(subs.get(id).timer);
          subs.delete(id);
        }

        const resource = variables.resource || 'events';
        const size = parseInt(variables.size, 10) || 1;
        const format = variables.format || 'json';
        const method = variables.method || 'GET';
        let intervalMs = parseInt(variables.intervalMs, 10) || 1000;
        if (intervalMs < MIN_INTERVAL_MS) intervalMs = MIN_INTERVAL_MS;

        const emit = () => {
          try {
            const data = makeMockStreamResponse(method, resource, size, format);
            send({ id, type: 'next', payload: { data: { mockStream: data } } });
          } catch (err) {
            send({
              id,
              type: 'error',
              payload: [{ message: err.message, validSizes: VALID_SIZES_KB }],
            });
            clearInterval(subs.get(id)?.timer);
            subs.delete(id);
          }
        };

        emit();
        const timer = setInterval(emit, intervalMs);
        subs.set(id, { timer });
        return;
      }

      if (type === 'complete') {
        if (subs.has(id)) {
          clearInterval(subs.get(id).timer);
          subs.delete(id);
        }
        send({ id, type: 'complete' });
        return;
      }

      if (type === 'ping') {
        send({ type: 'pong' });
      }
    });

    ws.on('close', () => {
      for (const { timer } of subs.values()) clearInterval(timer);
      subs.clear();
    });
  });

  return wss;
}

module.exports = { attachGraphqlWs };
